obstore/python/obstore/auth/earthdata.py

```diff
@@ -6,157 +6,419 @@
+                `None`, in which case, environment variables are used, if set.
+
+                Precedence is as follows:
+
+                1. Uses the specified value, if not `None`.
+                2. Uses the environment variable `EARTHDATA_TOKEN`, if set.
+                3. Uses the environment variables `EARTHDATA_USERNAME` and
+                   `EARTHDATA_PASSWORD`, if both are set.
+                4. Uses netrc to locate a username and password for `host`.
+                   Uses the environment variable `NETRC`, if set, to locate a
+                   netrc file; otherwise, uses the default netrc file location
+                   (`~/.netrc` on non-Windows OS or `~/_netrc` on Windows).
 
-        # Pass region default
+            session: The requests session to use for making requests to obtain S3
+                credentials. Defaults to `None`, in which case a default session
+                is created internally.  In this case, use this credential
+                provider's `close` method to release resources when you are
+                finished with it.
+
+        """
         self.config = {"region": "us-west-2"}
-        self.session = requests.Session()
-        self.session.auth = (username, password)
+        self._session: requests.Session | None = session or default_requests_session()
+        # Avoid closing a user-supplied session (the user is responsible for that)
+        self._close = None if session else self._session.close
+        self._get_credentials = _make_get_credentials(
+            credentials_url,
+            host or _default_host(),
+            auth or _default_auth(),
+        )
 
     def __call__(self) -> S3Credential:
         """Request updated credentials."""
-        resp = self.session.get(CREDENTIALS_API, allow_redirects=True, timeout=15)
-        auth_resp = self.session.get(resp.url, allow_redirects=True, timeout=15)
-        creds = auth_resp.json()
-        return {
-            "access_key_id": creds["accessKeyId"],
-            "secret_access_key": creds["secretAccessKey"],
-            "token": creds["sessionToken"],
-            "expires_at": datetime.fromisoformat(creds["expiration"]),
-        }
+        if self._session is None:
+            msg = "credential provider was closed"
+            raise ValueError(msg)
+
+        metadata = self._get_credentials(self._session)
+
+        return _metadata_to_s3_credential(metadata)
 
     def close(self) -> None:
         """Close the underlying session.
 
         You should call this method after you've finished all obstore calls to close the
-        underlying [requests.Session][].
+        underlying [requests.Session][].  If you supplied your own session object,
+        it will not be closed.
         """
-        self.session.close()
+        if self._close:
+            self._session = None
+            self._close()
+
+
+def _make_get_credentials(
```

Comment:
Ditto about the suggestion to make this a class instance method instead of a factory function.